napari/_qt/_qplugins/_qnpe2.py

```diff
@@ -0,0 +1,210 @@
+    _get_contrib_parent_menu,
+    get_widget_contribution,
+)
+from napari.utils.events import Event
+from napari.utils.translations import trans
+from napari.viewer import Viewer
+
+if TYPE_CHECKING:
+    from npe2.manifest import PluginManifest
+    from npe2.plugin_manager import PluginName
+    from npe2.types import WidgetCreator
+
+    from napari._qt.qt_main_window import Window
+
+
+def _get_widget_viewer_param(
+    widget_callable: WidgetCreator, widget_name: str
+) -> str:
+    """Get widget parameter name for `viewer` (if any) and check type."""
+    if inspect.isclass(widget_callable) and issubclass(
+        widget_callable,
+        (QWidget, Widget),
+    ):
+        widget_param = ""
+        # Inspection can fail when adding to bundle as it thinks widget is
+        # a builtin
+        try:
+            sig = inspect.signature(widget_callable.__init__)
+        except ValueError:
+            pass
+        else:
+            for param in sig.parameters.values():
+                if param.name == 'napari_viewer' or param.annotation in (
+                    'napari.viewer.Viewer',
+                    Viewer,
+                ):
+                    widget_param = param.name
+                    break
+
+    # For magicgui type widget contributions, `Viewer` injection is done by
+    # `magicgui.register_type`.
+    elif isinstance(widget_callable, MagicFactory) or inspect.isfunction(
+        widget_callable
+    ):
+        widget_param = ""
+    else:
+        raise TypeError(
+            trans._(
+                "'{widget}' must be `QtWidgets.QWidget` or `magicgui.widgets.Widget` subclass, `MagicFactory` instance or function. Please raise an issue in napari GitHub with the plugin and widget you were trying to use.",
+                deferred=True,
+                widget=widget_name,
+            )
+        )
+    return widget_param
+
+
+def _toggle_or_get_widget(
+    napari_viewer: Viewer,
+    mf: PluginManifest,
+    widget_name: str,
+    name: str,
+) -> Optional[Tuple[Union[FunctionGui, QWidget, Widget], str]]:
+    """Toggle if widget already built otherwise return widget.
+
+    Returned widget will be added to main window by a processor.
+    Note for magicgui type widget contributions, `Viewer` injection is done by
+    `magicgui.register_type` instead of a provider via annnotation.
+    """
+    window = napari_viewer.window
+    if dock_widget := window._dock_widgets.get(name):
+        dock_widget.setVisible(not dock_widget.isVisible())
+        return None
+
+    # Get widget param name (if any) and check type
+    if widget_contribution := get_widget_contribution(
+        mf.name,
+        widget_name,
+    ):
+        widget_callable, _ = widget_contribution
+        widget_param = _get_widget_viewer_param(widget_callable, widget_name)
+
+        kwargs = {}
+        if widget_param:
+            kwargs[widget_param] = napari_viewer
+        return widget_callable(**kwargs), name
+    return None
```

Comment:
Yes this can raise exception.

I did not know if there was some special reason for `get_widget_contribution` to return `None`, since you pass it a widget name (i.e., did you just make a widget that does not exist)? But this may just be defensive programming.

Reply:
I've used `RuntimeError` as was not sure of what to use. Can change.

Reply:
Okay after discussing with @DragaDoncila, `get_widget_contribution` should never return `None` when we are passing in a widget name. The `if` is just because of mypy. I can remove the `if` and add a mypy ignore comment instead.